### PR TITLE
Mtl h

### PR DESCRIPTION
**Pull request by mkopec**

### CODE REVIEW

File: src/soc/intel/meteorlake/meminit.c [hunk 1]
```diff
@@ -38,18 +38,16 @@ static const struct soc_mem_cfg soc_mem_cfg[] = {
 		.num_phys_channels = DDR5_CHANNELS,
 		.phys_to_mrc_map = {
 			[0] = 0,
-			[1] = 1,
```
Comment (krystian-hebel) [inline src/soc/intel/meteorlake/meminit.c:41]: While all changes in this file look reasonable, it breaks existing platforms, doesn't it?

File: src/soc/intel/meteorlake/romstage/fsp_params.c [hunk 2]
```diff
@@ -165,7 +167,7 @@ static void fill_fspm_mrc_params(FSP_M_CONFIG *m_cfg,
 	m_cfg->RMC = 0;
 	m_cfg->MarginLimitCheck = 0;
 	/* Enable MRC Fast Boot */
-	m_cfg->MrcFastBoot = 1;
+	m_cfg->MrcFastBoot = !vbnv_cmos_failed();
```
Comment (krystian-hebel) [inline src/soc/intel/meteorlake/romstage/fsp_params.c:170]: AFAICT this won't compile without `CONFIG_VBOOT`, and there is nothing in Kconfig for SoC that enforces it to be enabled. In any case, I don't think this is a good plate to add dependency on vboot to.

**Review by krystian-hebel — CHANGES_REQUESTED** [review]
Commits need to be cleaned up, [do many things](https://github.com/Dasharo/coreboot/pull/488/commits/caea8450a74f604a1616f8540b24f5604e651f4b) isn't a proper commit message.

File: src/soc/intel/meteorlake/romstage/fsp_params.c (hunk 2 above)
Comment (mkopec) [reply src/soc/intel/meteorlake/romstage/fsp_params.c:170]: It doesn't actually depend on Vboot. The function is defined in https://github.com/Dasharo/coreboot/blob/691bec1b71bdb814fef1f55fe175a19af6923aab/src/soc/intel/meteorlake/pmutil.c#L179-L182 and is compiled regardless if Vboot is enabled or not.

File: src/soc/intel/meteorlake/graphics.c [hunk 3]
```diff
@@ -0,0 +1,11 @@
+/* SPDX-License-Identifier: GPL-2.0-or-later */
+
+#include <intelblocks/graphics.h>
+#include <soc/ramstage.h>
+
+const struct i915_gpu_controller_info *
+intel_igd_get_controller_info(const struct device *const dev)
+{
+	const struct soc_intel_meteorlake_config *const chip = dev->chip_info;
+	return &chip->gfx;
```
[the remark this answers is not in the document]
Comment (mkopec) [reply src/soc/intel/meteorlake/graphics.c:10]: the null check is done by the caller: https://github.com/Dasharo/coreboot/blob/691bec1b71bdb814fef1f55fe175a19af6923aab/src/soc/intel/common/block/graphics/graphics.c#L179

do note that this function is basically identical for all platforms from skylake to alderlake

File: src/soc/intel/common/block/cse/cse_eop.c [hunk 4]
```diff
@@ -301,6 +301,11 @@ static bool is_cse_eop_supported(void)
 		       "SOFT TEMP DISABLE state, skipping EOP\n");
 		return false;
 	}
+	if (cse_is_hfs1_com_debug()) {
+		printk(BIOS_INFO, "HECI: coreboot in recovery mode; found CSE in expected "
```
[the remark this answers is not in the document]
Comment (mkopec) [reply src/soc/intel/common/block/cse/cse_eop.c:305]: Fair, but it's a copypaste of the line 3 lines above :)

File: src/soc/intel/common/block/smbus/smbuslib.c [hunk 5]
```diff
@@ -37,6 +40,61 @@ static void smbus_read_spd(u8 *spd, u8 addr)
 	}
 }
 
+static void switch_page(u8 spd_addr, u8 new_page)
+{
+	u32 offset;
+	/*
+	 *  By default,an SPD5 hub accepts 1 byte addressing pointing
```
[the remark this answers is not in the document]
Comment (mkopec) [reply src/soc/intel/common/block/smbus/smbuslib.c:47]: This is a cherrypick of https://review.coreboot.org/c/coreboot/+/52731 , maybe let's send those comments there?

File: src/soc/intel/common/block/smbus/smbuslib.c [hunk 6]
```diff
@@ -37,6 +40,61 @@ static void smbus_read_spd(u8 *spd, u8 addr)
 	}
 }
 
+static void switch_page(u8 spd_addr, u8 new_page)
+{
+	u32 offset;
```
[the remark this answers is not in the document]
Comment (mkopec) [reply src/soc/intel/common/block/smbus/smbuslib.c:45]: This is a cherrypick of https://review.coreboot.org/c/coreboot/+/52731 , maybe let's send those comments there?

File: src/soc/intel/common/block/cse/cse_eop.c (hunk 4 above)
[the remark this answers is not in the document]
Comment (mkopec) [reply src/soc/intel/common/block/cse/cse_eop.c]: Fixed all 3 instances in https://github.com/Dasharo/coreboot/pull/488/commits/d662158e798f1a73bdb6c97878c5116292802622

File: src/mainboard/clevo/mtl-h/ramstage.c [hunk 7]
```diff
@@ -0,0 +1,212 @@
...
+		break;
+	case FAN_CURVE_OPTION_SILENT:
+	default:
+		cmd.curve = fan_curve_silent;
+		break;
+	}
+
+	for (i = 0; i < (CONFIG(EC_SYSTEM76_EC_DGPU) ? 2 : 1); ++i) {
+		cmd.fan = i;
+		system76_ec_smfi_cmd(CMD_FAN_CURVE_SET, sizeof(cmd) / sizeof(uint8_t), (uint8_t *)&cmd);
```
[the remark this answers is not in the document]
Comment (mkopec) [reply src/mainboard/clevo/mtl-h/ramstage.c]: Fixed in commit 04ca4f5f25b53217b3e810e6fb5e753942021576

File: configs/config.novacustom_v540tu [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+# CONFIG_CONSOLE_POST is not set
+CONFIG_IFD_BIN_PATH="3rdparty/dasharo-blobs/novacustom/v5x0tu/descriptor.bin"
+CONFIG_ME_BIN_PATH="3rdparty/dasharo-blobs/novacustom/v5x0tu/me.bin"
+CONFIG_CONSOLE_CBMEM_BUFFER_SIZE=0x200000
+CONFIG_PCIEXP_DEFAULT_MAX_RESIZABLE_BAR_BITS=36
+CONFIG_HAVE_IFD_BIN=y
+CONFIG_EDK2_BOOTSPLASH_FILE="3rdparty/dasharo-blobs/novacustom/bootsplash.bmp"
+CONFIG_TPM_MEASURED_BOOT=y
+CONFIG_POWER_STATE_OFF_AFTER_FAILURE=y
+CONFIG_FSP_HEADER_PATH="3rdparty/blobs/mainboard/clevo/mtl-h/FSP/MeteorLake/Include/"
```
[the remark this answers is not in the document]
Comment (mkopec) [reply configs/config.novacustom_v540tu:17]: Should be fixed by https://github.com/Dasharo/coreboot/pull/488/commits/17e76bd70a04a32e56ddfe8a99fa9b4d9fec2500

File: configs/config.novacustom_v560tu [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+CONFIG_IFD_BIN_PATH="3rdparty/dasharo-blobs/novacustom/v5x0tu/descriptor.bin"
+CONFIG_ME_BIN_PATH="3rdparty/dasharo-blobs/novacustom/v5x0tu/me.bin"
+CONFIG_CONSOLE_CBMEM_BUFFER_SIZE=0x200000
+CONFIG_PCIEXP_DEFAULT_MAX_RESIZABLE_BAR_BITS=36
+CONFIG_HAVE_IFD_BIN=y
+CONFIG_EDK2_BOOTSPLASH_FILE="3rdparty/dasharo-blobs/novacustom/bootsplash.bmp"
+CONFIG_TPM_MEASURED_BOOT=y
+CONFIG_POWER_STATE_OFF_AFTER_FAILURE=y
+CONFIG_FSP_HEADER_PATH="3rdparty/blobs/mainboard/clevo/mtl-h/FSP/MeteorLake/Include/"
+CONFIG_FSP_FD_PATH="3rdparty/blobs/mainboard/clevo/mtl-h/FSP/MeteorLake/Fsp.fd"
```
[the remark this answers is not in the document]
Comment (mkopec) [reply configs/config.novacustom_v560tu:19]: Should be fixed by https://github.com/Dasharo/coreboot/pull/488/commits/17e76bd70a04a32e56ddfe8a99fa9b4d9fec2500

File: configs/config.novacustom_v540tu (hunk 8 above)
[the remark this answers is not in the document]
Comment (mkopec) [reply configs/config.novacustom_v540tu:17]: Also needed https://github.com/Dasharo/coreboot/pull/488/commits/ea23d1523b9a3158a40601630330169027bcb992 , now CI passes.

File: src/soc/intel/common/block/smbus/smbuslib.c (hunk 5 above)
[the remark this answers is not in the document]
Comment (mkopec) [reply src/soc/intel/common/block/smbus/smbuslib.c:47]: Also needed https://github.com/Dasharo/coreboot/pull/488/commits/ea23d1523b9a3158a40601630330169027bcb992 , now CI passes.

**Comment by mkopec** [comment]
@krystian-hebel Squashed all the garbage commits and dropped unneeded changes: https://github.com/Dasharo/coreboot/pull/488/commits/94856afa0fcdf61bf709704ff80a581be20826d9 

Commit count went from 113 to 41

File: configs/config.novacustom_v560tu (hunk 9 above)
[the remark this answers is not in the document]
Comment (mkopec) [reply configs/config.novacustom_v560tu:19]: Also needed https://github.com/Dasharo/coreboot/pull/488/commits/ea23d1523b9a3158a40601630330169027bcb992 , now CI passes.

File: src/soc/intel/meteorlake/meminit.c (hunk 1 above)
Comment (mkopec) [reply src/soc/intel/meteorlake/meminit.c:41]: I'm not sure. A similar point was raised upstream too: https://review.coreboot.org/c/coreboot/+/82733/1

File: src/soc/intel/common/block/smbus/smbuslib.c (hunk 6 above)
[the remark this answers is not in the document]
Comment (krystian-hebel) [reply src/soc/intel/common/block/smbus/smbuslib.c:45]: https://review.coreboot.org/c/coreboot/+/52731/16#message-73a59c3375d77c44bf2420318297812a14c7e9c3

File: src/soc/intel/common/block/smbus/smbuslib.c [hunk 10]
```diff
@@ -52,22 +110,31 @@ static int get_spd(u8 *spd, u8 addr)
...
 
-		if (i2c_eeprom_read(addr, 0, SPD_PAGE_LEN, spd + SPD_PAGE_LEN) < 0) {
+		if (i2c_eeprom_read(addr, 0, SPD_PAGE_LEN, spd) < 0) {
 			printk(BIOS_INFO, "do_i2c_eeprom_read failed, using fallback\n");
-			smbus_read_spd(spd + SPD_PAGE_LEN, addr);
+			smbus_read_spd(spd, addr);
+		}
+
+		/* Check if module is DDR4, DDR4 spd is 512 byte. */
+		if (spd[SPD_DRAM_TYPE] == SPD_DRAM_DDR4 &&
```
[the remark this answers is not in the document]
Comment (krystian-hebel) [reply src/soc/intel/common/block/smbus/smbuslib.c:126]: https://review.coreboot.org/c/coreboot/+/52731/16#message-73a59c3375d77c44bf2420318297812a14c7e9c3

File: src/soc/intel/meteorlake/romstage/fsp_params.c (hunk 2 above)
Comment (krystian-hebel) [reply src/soc/intel/meteorlake/romstage/fsp_params.c:170]: Right, I only checked that it is declared in  https://github.com/Dasharo/coreboot/blob/dasharo/src/security/vboot/vbnv.h#L27, my bad.

**Comment by krystian-hebel** [comment]
> @krystian-hebel Squashed all the garbage commits and dropped unneeded changes: [94856af](https://github.com/Dasharo/coreboot/commit/94856afa0fcdf61bf709704ff80a581be20826d9)
> 
> Commit count went from 113 to 41

Thanks, looks much better. Another rebase will be needed becvause of https://github.com/Dasharo/coreboot/pull/559, it should fic CI issues, but I don't expect too many conflicts.

File: src/soc/intel/common/block/smbus/smbuslib.c (hunk 10 above)
[the remark this answers is not in the document]
Comment (mkopec) [reply src/soc/intel/common/block/smbus/smbuslib.c:126]: @miczyg1 could you help here?

File: src/soc/intel/common/block/smbus/smbuslib.c (hunk 6 above)
[the remark this answers is not in the document]
Comment (mkopec) [reply src/soc/intel/common/block/smbus/smbuslib.c:45]: @miczyg1 could you help here?

File: src/soc/intel/common/block/smbus/smbuslib.c (hunk 5 above)
[the remark this answers is not in the document]
Comment (mkopec) [reply src/soc/intel/common/block/smbus/smbuslib.c:47]: @miczyg1 could you help here?

File: src/soc/intel/common/block/smbus/smbuslib.c (hunk 10 above)
[the remark this answers is not in the document]
Comment (miczyg1) [reply src/soc/intel/common/block/smbus/smbuslib.c:126]: How may I be of help here? It is just refactoring for better readability and maintainability of the code.

File: src/lib/spd_cache.c [hunk 11]
```diff
@@ -209,7 +209,9 @@ enum cb_err spd_fill_from_cache(uint8_t *spd_cache, struct spd_block *blk)
 
 	dram_type = *(spd_cache + SC_SPD_OFFSET(i) + SPD_DRAM_TYPE);
 
-	if (dram_type == SPD_DRAM_DDR4)
+	if (dram_type == SPD_DRAM_DDR5)
+		blk->len = CONFIG_DIMM_SPD_SIZE;
```
Comment (miczyg1) [inline src/lib/spd_cache.c:213]: Shouldn't it be something like SPD_PAGE_LEN_DDR5?

File: src/soc/intel/common/block/smbus/smbuslib.c (hunk 6 above)
[the remark this answers is not in the document]
Comment (mkopec) [reply src/soc/intel/common/block/smbus/smbuslib.c]: renamed in commit c812b8e493b381fa8b4f4bd92cafc441e4a17326

File: src/soc/intel/common/block/smbus/smbuslib.c (hunk 5 above)
[the remark this answers is not in the document]
Comment (mkopec) [reply src/soc/intel/common/block/smbus/smbuslib.c]: applied in c812b8e493b381fa8b4f4bd92cafc441e4a17326

File: src/soc/intel/common/block/smbus/smbuslib.c (hunk 10 above)
[the remark this answers is not in the document]
Comment (mkopec) [reply src/soc/intel/common/block/smbus/smbuslib.c]: refactored in 5ee9aa04a9fbf1f3f538bf25d94b0f939d855e11

File: src/lib/spd_cache.c (hunk 11 above)
Comment (mkopec) [reply src/lib/spd_cache.c]: indeed, done in ba0140a9076a77760d1cb5e0a94db6e939abe726

**Comment by mkopec** [comment]
@miczyg1 @krystian-hebel I've addressed both of yours' comments

File: src/lib/spd_cache.c (hunk 11 above)
Comment (krystian-hebel) [reply src/lib/spd_cache.c]: I disagree on the naming (but not on the fact that it shouldn't be `CONFIG_DIMM_SPD_SIZE`), now `MAX_SPD_PAGE_SIZE_SPD5` < `SPD_PAGE_LEN_DDR5`.

Comment (mkopec) [reply src/lib/spd_cache.c]: Renamed to SPD_LEN_DDR5 in https://github.com/Dasharo/coreboot/compare/ba0140a9076a77760d1cb5e0a94db6e939abe726..bd6336a138437d57fdc4c2097a664150e417d83f

**Review by krystian-hebel — APPROVED** [review]
Approved, but needs a rebase.

**Comment by miczyg1** [comment]
TBH I can no longer see if there were any of my comments here. After rebasing it can land.

**Review by miczyg1 — APPROVED** [review]

**Review by miczyg1 — APPROVED** [review]

**Comment by miczyg1** [comment]
@mkopec for some reasons oen of your changes breaks CI check for ODROID config file... When I do:

```
coreboot@2a04f4fb2c50:~/coreboot$ make distclean
coreboot@2a04f4fb2c50:~/coreboot$ cp configs/config.hardkernel_odroid_h4 .config
coreboot@2a04f4fb2c50:~/coreboot$ make olddefconfig
#
# configuration written to /home/coreboot/coreboot/.config
#
coreboot@2a04f4fb2c50:~/coreboot$ make savedefconfig
coreboot@2a04f4fb2c50:~/coreboot$ cp defconfig configs/config.hardkernel_odroid_h4 
coreboot@2a04f4fb2c50:~/coreboot$ git diff
```

I get the same results as on CI using mtl-h branch. But on dasharo branch it doesn't happen...

**Pull request merged**